Steps blending v1

dnerini opened the pull request:
This PR needs to implement the changes needed to make the blending module compatible with v1 (i.e. without xarray).

RubenImhoff committed:
Implement STEPS blending (#233)

* First basic functions to implement STEPS blending

* Add compute of blend means,sigmas and recompose

* pysteps.io with xarray (#219)

* Add xarray dependency

* MCH importer returns an xarray Dataset

* Remove plot lines

* Remove import

* Adapt readers to xarray format

* Rewrite as more general decorator

* Add missing metadata

* Adapt io tests

* Mrms bounding box (#222)

* Fix bounding box coordinates

* Add missing metadata

* Import xarray DataArray

Ignore quality field

* Black

* Do not hardcode metadata

* Address review comments by ruben

* Add a legacy option to the io functions

A "legacy" options is added to revert back the importers and readers behavior to version 1. This is a temporary solution to allow the examples, and other functions, to run as usual (v1.*).

Hopefully, this is will allow a easier transition into version 2 during the development process and will allow testing functions that were not updated to v2.

* Fix compatibility problems with tests

Many of the tests were updated to use the legacy data structures (v1). The tests that still contains issues, were tagged with a TODO message and they are skipped.

This will allow incremental changes to be tested in the new v2.

IMPORTANT: once the v2 branch is stable, we may remove the legacy compatibility from the code base and the tests.

* Update dependencies

* Ignore plugins test

Co-authored-by: Andres Perez Hortal <16256571+aperezhortal@users.noreply.github.com>

* Add blend_optical_flow

* changes to steps blending procedure - weights according to adjusted BPS2006 method

* changes to blending procedures - adjust weights from original BPS2006 method

* Determine spatial correlation of NWP model forecast

* First attempt to make correlations and thus weights lead time dependent (in progress..)

* Change back to original BPS2006 blending formulation and add regression of skill values to climatological values for weights determination

* Reformat code with Black

* Skill score script imports climatological correlation-values from file now

* Small changes to skill score script

* Add skill score tests and an interface

* Add skill score tests and an interface

* Small change to docstring

* Bom import xarray (#228)

* Add import_bom_rf3  using xarray

* Add tests to xarray version

* Fix mrms importer tests

* Pass **kwargs to internal functions

* Add nwp_importers to read bom nwp sample data

* Add bom nwp data to source file

* Add tests for bom_nwp reader

* Fix pystepsrc

Co-authored-by: Andres Perez Hortal <16256571+aperezhortal@users.noreply.github.com>

* Functions to store and compute climatological weights (#231)

* Implement the functions get_default_weights, save_weights, calc_clim_weights. 

These functions are used to evolve the weights in the scale- and skill-dependent blending with NWP in the STEPS blending algorithm. The current weights, based on the correlations per cascade level, are regressed towards these climatological weights in the course of the forecast.

These functions save the current and compute the climatological weights (a running mean of the weights of the past n days, where typically n=30). First daily averages are stored and these are then averaged over the running window of n days.

* Add tests for pysteps climatological weight io and calculations.

* Add path_workdir to outputs section in pystepsrc file and use it as a default path to store/retrieve blending weights.

* Minor changes to docstrings, changes to skill scores and testing scripts

* Completed documentation for blending clim module, cleanup.

Co-authored-by: RubenImhoff <r.o.imhoff@live.nl>

* Main blending module, first steps

* Add simple tests

* Minor changes to tester: velocity now based on rainfall field of NWP

* Add utilities to decompose, store and load NWP cascades for use in blending  (#232)

* First version of NWP decomposition

* Added saving to netCDF

* Completed functions for saving and loading decomposed NWP data

* Added example files for the decomposed NWP functions

* Added compatibility with numpy datetime

* Use default output path_workdir for tmp files in blending/utils.py.

* Update documentation of NWP decomposition functions in utils.py

Co-authored-by: Wout Dewettinck <wout.dewettinck@ugent.be>
Co-authored-by: wdewettin <87696913+wdewettin@users.noreply.github.com>

* Add importer for RMI NWP data (#234)

Add importer for netcdf NWP data from RMI using xarrays.

* Add test for RMI NWP data importer.

* Add entry for RMI NWP data in pystepsrc.

* Run black on everything: fix formatting.

* Add KNMI Harmonie NWP netcdf importer and tests (#235)

* Changes to v_models to make use of multiple timesteps. Changes in the velocity field over time in the NWP forecast will be taken into account now.

* Fixes for KNMI importer:

Add forgotten @postprocess_import()
Don't call dropna on NWP data.

* Avoid shadowing of pysteps.blending.utils by pysteps.utils

* First attempt for probability matching and masking utility; part 1

* Changes to prob matching and masking methods; part 2

* Prob matching and masking changes; part 3. Ready for testing with real data from here on

* Remove unnecessary print statements

* Cleanup imports

* More cleanup

* Update docstrings

* RMI importer for gallery example (will follow)

* Reprojection functionality (#236)

* Added Lesley's reprojection module to this branch

* Added compatibility for three-dimensional xarrays

* Add commentary to reprojection util

* Changes to make reprojection of KNMI data possible

* Changes after Daniele's review

* Add dependencies

* Changes to importers, see issue #215

* Add tests

* Fix some issues

* documentation

* Fixes for tests

* Set requirements again

* Some fixes

* Changes to nwp_importers after Carlos' response

* Remove wrong example script

* Remove rasterio dependencies from lists

* First try to prevent testing error

* Changes Daniele and fix knmi nwp importer

* Add rasterio to tox.ini

* Aesthetics

* rasterio import test

* Add rasterio to the test dependencies

* Reset try-except functionality for rasterio import

* Fix for failing test on windows python 3.6

* add importerskip rasterio

Co-authored-by: Wout Dewettinck <wout.dewettinck@ugent.be>

* Fixes in nwp importers

* Revert "Merge branch 'steps_blending' into pysteps-v2" (#239)

This reverts commit 2c639f80e71a766510b41c6190bc1b23d9691d8a, reversing
changes made to bccb8fc2e8fd939906d070443d27cf0ad1d2d70d.

* Merge latest version pysteps-v2 into steps_blending branch (#237)

* Update docstrings

* More cleanup

* Cleanup imports

* Cleanup imports

* More cleanup

* Update docstrings

* Update references

Mention the work of Ravuri et al (2021, Nature) as an example of work using cGANs to generate ensembles

* Clean up page

* Reprojection functionality (#236)

* Added Lesley's reprojection module to this branch

* Added compatibility for three-dimensional xarrays

* Add commentary to reprojection util

* Changes to make reprojection of KNMI data possible

* Changes after Daniele's review

* Add dependencies

* Changes to importers, see issue #215

* Add tests

* Fix some issues

* documentation

* Fixes for tests

* Set requirements again

* Some fixes

* Changes to nwp_importers after Carlos' response

* Remove wrong example script

* Remove rasterio dependencies from lists

* First try to prevent testing error

* Changes Daniele and fix knmi nwp importer

* Add rasterio to tox.ini

* Aesthetics

* rasterio import test

* Add rasterio to the test dependencies

* Reset try-except functionality for rasterio import

* Fix for failing test on windows python 3.6

* add importerskip rasterio

Co-authored-by: Wout Dewettinck <wout.dewettinck@ugent.be>

* Revert "Merge branch 'steps_blending' into pysteps-v2" (#239)

This reverts commit 2c639f80e71a766510b41c6190bc1b23d9691d8a, reversing
changes made to bccb8fc2e8fd939906d070443d27cf0ad1d2d70d.

Co-authored-by: ned <daniele.nerini@meteoswiss.ch>
Co-authored-by: dnerini <daniele.nerini@gmail.com>
Co-authored-by: Wout Dewettinck <wout.dewettinck@ugent.be>

* NWP skill calculation only within radar domain

* Update docs

* Add example for gallery examples

* Fix docstrings example

* Remove additional normalization step

* Fixes for the tests

* update docs

* changes to post-processing rainfall field and docstrings

* Update contributing guidelines (#241)

- Improve grammar.
- Make the guide more concise. Remove unused/unnecessary rules.
- Indicate more clearly which parts of the guidelines are inspired by other projects (before they were only mentioned at the end).
- Change "Travis-CI" references by "GitHub Actions".

* Advect noise cascade

* Allow for moving domain mask of extrapolation component

* minor fixes

* Linear blending (#229)

* Implemented linear blending function
* Added example file and test
* Added compatibility for NWP ensembles

The PR is ready to go. Making the code xarray ready will be done in a separate PR. 

Co-authored-by: RubenImhoff <r.o.imhoff@live.nl>

* weights calculation adjustment outside radar domain if only one model present

* allow for mirroring of advected noise cascade

* implementation of weights following Seed et al. (2013)

* Allow for decomposed NWP precip and NWP velocity fields: part 2

* Store decomposed fields with compression

* changes after first review Daniele

* Remove unnecessary print statement

* fixes to blending utils and implementation of blending utils tests

* remove unnecessary lines

* Fix one time step shift of extrapolation skill prior to blending

* minor changes to blending climatology, blending weights and remove path_workdir from pystepsrc

* Make NWP forecast decomposition prior to blending function optional

* Use pathlib

* Extract methods

* Minor changes to docstrings

* Access climatological skill file for multiple NWP model and date string changes to prevent errors in blending.utils

Co-authored-by: Carlos Velasco <carlos.velasco@bom.gov.au>
Co-authored-by: ned <daniele.nerini@meteoswiss.ch>
Co-authored-by: Andres Perez Hortal <16256571+aperezhortal@users.noreply.github.com>
Co-authored-by: Ruben Imhoff <Ruben.Imhoff@deltares.nl>
Co-authored-by: Carlos Velasco <cvelascof@gmail.com>
Co-authored-by: Lesley De Cruz <lesley.decruz+git@gmail.com>
Co-authored-by: Wout Dewettinck <wout.dewettinck@ugent.be>
Co-authored-by: wdewettin <87696913+wdewettin@users.noreply.github.com>
Co-authored-by: Lesley De Cruz <lesley.decruz@meteo.be>
Co-authored-by: dnerini <daniele.nerini@gmail.com>

dnerini committed:
First changes to remove xarray dependency

dnerini commented:
In order to merge into master v1 there are few pending issues that we need to solve:

## 1. Remove xarray hard dependency
I already started addressing this with b947b6b but some work remains. In particular, the reprojection module is based on xarray and needs to be migrated to numpy. Alternatively, we could make xarray an optional dependency. I also suggest to remove the bom importers based on xarray.
@ladc @RubenImhoff  @cvelascof 

## 2. NWP Importers
NWP importers should be separated from the base package as plugins, see #218. This could also a way to address the issue above, since the xarray dependency could be moved to the external plugin. For now, we suggest we do the exercise with the KNMI NWP importer and use that one for testing and examples of blending in the base package. What do you think @RubenImhoff?

RubenImhoff commented:
I'll have a look at it (in the coming days) too!

dnerini committed:
Do not format json with black

dnerini committed:
Comment out xarray code

dnerini committed:
Remove 'legacy' argument

dnerini committed:
Fix black

RubenImhoff commented:
@ladc, I see that there is an `examples/test_decomposition_NWP.py`. Do we want to keep this example? It could be useful, I think, but perhaps we should add some more doc strings and printed output in it.

RubenImhoff committed:
Fix issues from codacy report - v1

RubenImhoff committed:
Fix issues from codacy report - v2

dnerini committed:
Set minimum python version to 3.7  (#253)

* Set minimum python version to 3.7

and maximum version to 3.9

* Update github actions

Co-authored-by: Andres Perez Hortal <16256571+aperezhortal@users.noreply.github.com>

dnerini committed:
Ignore vscode and macos folders

RubenImhoff commented:
@dnerini, how should we move forward with the external nwp importer plugins? Do you want to use the cookiecutter implementation or make a separate plugin that basically uses the current `nwp_importers.py` implementation?

RubenImhoff committed:
Change reprojection, test and example script to work without xarray

RubenImhoff commented:
I changed the reprojection script to work without xarray. It (for now) assumes that `nwp_importers` returns an array with the rainfall fields and a separate metadata dictionary (similar to `io.importers.py`). So the reprojection tests should pass if these importers return that.

dnerini commented:
> hould we move forward with the external nwp importer plugins? Do you want to use the cookiecutter implementation or make a separate plugin that basically uses the current `nwp_importers.py` implementation?

yes I think we should give the cookicutter plugin from @aperezhortal a try to implement the nwp importers! No idea if this will work ... can we start with the KNMI importer?

dnerini committed:
Merge branch 'master' into blending-v1

dnerini committed:
Fix test

RubenImhoff commented:
The cookiecutter is a nice tool, @aperezhortal! It took some time to get familiar with it, but I have all three NWP importers in a separate plugin now. A couple of questions to move forward (and let you review it):

- Should I upload the plugin as a separate repository in pySTEPS? Then you could review the PR there.
- How to link the plugin to pysteps (locally it is easy with the installer, but does this also directly work in a repository)?
- For the tests in the plugin, how to link the pysteps-data-master to it (similar to how that is done in pysteps)?

aperezhortal commented:
Hi @RubenImhoff, you wrote the first-ever known plugin :muscle: 

About your questions:

> Should I upload the plugin as a separate repository in pySTEPS? Then you could review the PR there.

That is good idea.  You can perhaps upload it to an "initial_commit" branch and then submit a PR to the main branch with an initial commit that is just a README. I will be happy to take a look at it.

> How to link the plugin to pysteps (locally it is easy with the installer, but does this also directly work in a repository)?

Do you mean to link the plugin to pysteps repository's workflow to load the NWP test data? If so, I can think of two ways of doing that. One is to add a new step to the CI workflow that installs the plugins after pysteps is installed. This seems like the simpler approach to me. For free, we are also testing the integration of this plugin.  
The other option is to avoid any explicit dependency on the plugin and create fixtures for tests in formats that are easy to read (like numpy of h5).  The advantage here is that the CI workflow does not depend on the plugin to run the test, which is also good since the latter is not an explicit dependency of the blending module. This will perhaps help to keep things simple. 
What do you think ? 

> For the tests in the plugin, how to link the pysteps-data-master to it (similar to how that is done in pysteps)? 

If only a few files are needed, one option is to create a small script that downloads those files before running the tests.

RubenImhoff commented:
Thanks for your quick response, @aperezhortal! I'll make a PR of the plugin - then we can also discuss further there. 

> You can perhaps upload it to an "initial_commit" branch and then submit a PR to the main branch with an initial commit that is just a README.

What do  you mean with the initial commit just containing a README? You mean just putting the README in there without the rest of the files? 

> Do you mean to link the plugin to pysteps repository's workflow to load the NWP test data? If so, I can think of two ways of doing that. One is to add a new step to the CI workflow that installs the plugins after pysteps is installed. This seems like the simpler approach to me. For free, we are also testing the integration of this plugin.
The other option is to avoid any explicit dependency on the plugin and create fixtures for tests in formats that are easy to read (like numpy of h5). The advantage here is that the CI workflow does not depend on the plugin to run the test, which is also good since the latter is not an explicit dependency of the blending module. This will perhaps help to keep things simple.
What do you think ?

The first option is easy with the examples script, but I believe no other tests (except for the previous `nwp_importers`) depend on the presence of the importers. As the examples gallery is being decoupled from the main repository anyway, the latter option might work as well (though the examples gallery repo requires the plugin then..). 

> If only a few files are needed, one option is to create a small script that downloads those files before running the tests.

Good idea, let's put it in a PR and continue from there.

aperezhortal commented:
> What do you mean with the initial commit just containing a README? You mean just putting the README in there without the rest of the files?

Yes. I mentioned that to make sure that the destination branch of the PR exists and leave it empty to avoid having work-in-progress code in the main branch.

RubenImhoff committed:
remove nwp_importers from main pysteps repo (available in separate plugin now)

RubenImhoff committed:
remove nwp_importer tests

RubenImhoff committed:
make reprojection tests independent of nwp importer

RubenImhoff committed:
fixes to reprojection

RubenImhoff committed:
Remove BoM xarray importer and make BoM importer tests independent of xarray

RubenImhoff committed:
Make importer tests xarray independent

RubenImhoff committed:
fixes from codacy report

RubenImhoff committed:
fix tests and add multiple model tests for linear blending module

RubenImhoff commented:
16 reported issues from the Codacy report left. Most are very minor, but not sure how to fix those. 

edit: down to 14 :)

RubenImhoff committed:
minor fix for codacy report

dnerini commented:
Excellent work @RubenImhoff !! I think we are very close to merging: tests are passing and the coverage delta is positive. Probably we can just ignore the remaining issues from codacy, I'll have a look.

Let give this a final look and see if everything is ready.

RubenImhoff commented:
Can I ask you to have a final look on the prob matching and masking implementation? Or did you already do that?

RubenImhoff committed:
add extra linear blending tests

RubenImhoff committed:
add blending interface tests

RubenImhoff committed:
minor fix to io_readers test

dnerini committed:
Remove private methods from docs

dnerini commented:
@RubenImhoff can you briefly explain how you're dealing with the external dependencies for the nwp importers? I assume we still need them for the examples and the tests?

RubenImhoff commented:
> can you briefly explain how you're dealing with the external dependencies for the nwp importers? I assume we still need them for the examples and the tests?

We don't need them anymore for the tests, luckily. I hard coded some metadata and made dummy data instead of importing the NWP data. 
The examples is a good point that I forgot to mention. For the `blended_forecast` (I believe that is the only one - I'll check), it still depends on the NWP importers. Since we are moving the examples to a separata repo (right?), we should think about how to incorporate that.

dnerini commented:
> > can you briefly explain how you're dealing with the external dependencies for the nwp importers? I assume we still need them for the examples and the tests?
> 
> We don't need them anymore for the tests, luckily. I hard coded some metadata and made dummy data instead of importing the NWP data. The examples is a good point that I forgot to mention. For the `blended_forecast` (I believe that is the only one - I'll check), it still depends on the NWP importers. Since we are moving the examples to a separata repo (right?), we should think about how to incorporate that.

Ok, very good, thanks! I like the approach with the dummy input data, makes testing much slimmer, well done!

I don't see problems for having an additional external dependency for the examples. We just need to explain it at the beginning of the tutorial and add the importers plugin to the list of doc requirements once it is available on pypi.

RubenImhoff commented:
> I don't see problems for having an additional external dependency for the examples. We just need to explain it at the beginning of the tutorial and add the importers plugin to the list of doc requirements once it is available on pypi.

Agree, that would work! Plan B would be to work with dummy data in the example too, but that wouldn't look as nice.

dnerini committed:
Install plugins for building examples

dnerini commented:
> > I don't see problems for having an additional external dependency for the examples. We just need to explain it at the beginning of the tutorial and add the importers plugin to the list of doc requirements once it is available on pypi.
> 
> Agree, that would work! Plan B would be to work with dummy data in the example too, but that wouldn't look as nice.

done in 146c6d5, have a look and change as you please ;-)

edit: I've activated the relative RTD built, but it'll succeed only when the pysteps-nwp-importers package is available on pypi

dnerini committed:
Install pysteps-nwp-importers from source

dnerini commented:
@RubenImhoff @ladc currently the `test_decomposition_NWP.py` script in the examples folder cannot be rendered in the example gallery because is missing the necessary formatting (mostly the docstrings at the top of the file). 

Do you want to include it or should we simply remove it from the folder?

dnerini commented:
Also, @RubenImhoff I'm getting a `ValueError: R_models does not contain sufficient lead times for this forecast` from the blending example when building the documentation:

```
Traceback (most recent call last):
  File "/Users/daniele/Dropbox/Repos/pysteps/examples/blended_forecast.py", line 154, in <module>
    precip_forecast = blending.steps.forecast(
  File "/Users/daniele/Dropbox/Repos/pysteps/pysteps/blending/steps.py", line 349, in forecast
    _check_inputs(R, R_d_models, V, V_models, timesteps, ar_order)
  File "/Users/daniele/Dropbox/Repos/pysteps/pysteps/blending/steps.py", line 1529, in _check_inputs
    raise ValueError(
ValueError: R_models does not contain sufficient lead times for this forecast
```

Any idea what the cause could be?

dnerini committed:
Fix branch name

dnerini committed:
Ignore NWP decomposition example for now

dnerini committed:
Remove xarray notation

RubenImhoff commented:
> Do you want to include it or should we simply remove it from the folder?

Removing it from the folder is good with me!

RubenImhoff commented:
> Also, @RubenImhoff I'm getting a `ValueError: R_models does not contain sufficient lead times for this forecast` from the blending example when building the documentation:
> 
> ```
> Traceback (most recent call last):
>   File "/Users/daniele/Dropbox/Repos/pysteps/examples/blended_forecast.py", line 154, in <module>
>     precip_forecast = blending.steps.forecast(
>   File "/Users/daniele/Dropbox/Repos/pysteps/pysteps/blending/steps.py", line 349, in forecast
>     _check_inputs(R, R_d_models, V, V_models, timesteps, ar_order)
>   File "/Users/daniele/Dropbox/Repos/pysteps/pysteps/blending/steps.py", line 1529, in _check_inputs
>     raise ValueError(
> ValueError: R_models does not contain sufficient lead times for this forecast
> ```
> 
> Any idea what the cause could be?

I'll have a look at it.

RubenImhoff committed:
update to blending example

RubenImhoff commented:
I think I found the problem: the NWP forecast was not correctly indexed (the end index was missing). I added it, but couldn't test it myself, due to still the same netcdf problem when reading the bom data (still no clue why and why only with the bom data..). Could you maybe test if the example runs locally?

Anyway, I also changed the file path and name towards the NWP forecast, because the NWP names are no longer in pystepsrc. It's a hard coded path now, which is probably not ideal. How should we read this with pysteps?

RubenImhoff committed:
remove os import from blended_forecast example

RubenImhoff committed:
remove test_decomposition from examples

RubenImhoff commented:
The Black test seems to get stuck at the jupyter notebook example, should we exclude it from the test somewhere?

dnerini committed:
Fix black

dnerini commented:
> I think I found the problem: the NWP forecast was not correctly indexed (the end index was missing). I added it, but couldn't test it myself, due to still the same netcdf problem when reading the bom data (still no clue why and why only with the bom data..). Could you maybe test if the example runs locally?

It worked for me!

> Anyway, I also changed the file path and name towards the NWP forecast, because the NWP names are no longer in pystepsrc. It's a hard coded path now, which is probably not ideal. How should we read this with pysteps?

Right, good point. I need to think more about this... (also the doc built is failing because the hardcoded path is not correct)

edit: the more I look at it, the more I think the "data_sources" section of the pystepsrc should be hosted in the pysteps-data repo...

> The Black test seems to get stuck at the jupyter notebook example, should we exclude it from the test somewhere?

Turned out it was something else causing it to fail (see  5cd752b ).

dnerini committed:
Minor doc fixes

dnerini committed:
Reintroduce nwp sources

dnerini committed:
Fix enum list in docstrings

RubenImhoff commented:
Got the blended_forecast example also working locally again - everything seems to work fine now. Looks like we're getting there! What steps are left, just a final review?

dnerini commented:
> Got the blended_forecast example also working locally again - everything seems to work fine now. Looks like we're getting there! What steps are left, just a final review?

There is a problem with the RTD builds that I need to investigate. It doesn't appear to be linked to this branch, but it would be anyway good to check how the new documentation looks like before merging.

Also maybe we can work a bit the linear blending example? As it is now it feels a bit too little verbose and too much code to be a tutorial imho. Also, maybe we can replace the dummy nwp with actual data?

Otherwise yes, we could do a final round of review, trying to involve a larger audience if possible? @ladc @cvelascof @loforest  @aitaten

RubenImhoff commented:
I can have a look at the linear blending example - I'll make it a little more comparable to the blended_forecast example

RubenImhoff committed:
Updated linear blending example

dnerini committed:
Do not build pdf

dnerini commented:
doc is building, here: https://pysteps.readthedocs.io/en/blending-v1

RubenImhoff committed:
fixes to figures sizes in blending examples

RubenImhoff committed:
Merge branch 'blending-v1' of https://github.com/pySTEPS/pysteps into blending-v1_RI

RubenImhoff committed:
update linear blending doc strings

RubenImhoff committed:
fixes from codacy report

RubenImhoff commented:
Hmm codacy was happy with me this time ;)

dnerini committed:
Merge branch 'master' into blending-v1

dnerini commented:
Personally I think this is ready to be merged!

dnerini committed:
Fix table

dnerini committed:
Plot results as columns

dnerini committed:
Improve docstrings

dnerini committed:
Fix black

dnerini committed:
Merge branch 'master' into blending-v1

dnerini committed:
Rename (some) variables

dnerini committed:
Merge branch 'master' into blending-v1

RubenImhoff commented:
Nice work with the variable names, @dnerini! That was was still somewhere on my to-do list, so great it's done. :)

RubenImhoff committed:
Make sure NWP correlations are always finite numbers

RubenImhoff committed:
Make sure climatological skill is always positive and finite

dnerini committed:
Fix reference in docstrings

dnerini committed:
List rasterio as optional dependency

dnerini committed:
Make n_ens_members a mandatory argument

dnerini committed:
Remove default parameters

dnerini committed:
Fix black

dnerini commented on pysteps/tests/test_utils_reprojection.py:
```diff
@@ -0,0 +1,113 @@
+
+pytest.importorskip("rasterio")
+
+root_path_radar = pysteps.rcparams.data_sources["rmi"]["root_path"]
```
The "rmi" entry is missing from pystepsrc, @ladc @RubenImhoff

dnerini replied:
my bad! I was looking at the wrong branch ...

dnerini committed:
Rename reprojection method

And various fixes to the docs

dnerini committed:
Add missing test

dnerini committed:
Use micromamba 0.20

RubenImhoff commented:
In the original STEPS paper and technical note (in particular, the BPS2004 technical note), the lag-1 and lag-2 parameters are allowed to regress to their climatological values (eq. 12 - 19). By doing so, the Phi parameters change over time, which enhances the AR process. I have not implemented this (yet) in the steps blending module and saw that it also isn't part of the pysteps steps nowcasting module (right?). Is it worth implementing this, or is it left out of pysteps for a reason? 

Other than that, this PR is good to go! :)

dnerini commented:
> In the original STEPS paper and technical note (in particular, the BPS2004 technical note), the lag-1 and lag-2 parameters are allowed to regress to their climatological values (eq. 12 - 19). By doing so, the Phi parameters change over time, which enhances the AR process. I have not implemented this (yet) in the steps blending module and saw that it also isn't part of the pysteps steps nowcasting module (right?). Is it worth implementing this, or is it left out of pysteps for a reason?
> 
> Other than that, this PR is good to go! :)

good that you mention it (and perhaps it could be included in the notes section of the method itself), but I'd leave this as "future development" ;-)

RubenImhoff committed:
add notes about lag-1 and lag-2 parameter regression

RubenImhoff commented:
As discussed with @dnerini, the PR is ready to be merged. Any further developments can go through separate PRs.
